src/azul/service/manifest_service.py

```diff
@@ -705,10 +707,68 @@ def write_to(self, output: IO[str]) -> Optional[str]:
             url = furl(config.service_endpoint(),
                        path=f'/repository/files/{uuid}',
                        args=dict(version=version, catalog=self.catalog))
+            name = self._sanitize_path(name)
             output.write(f'url={self._option(url.url)}\n'
                          f'output={self._option(name)}\n\n')
         return None
 
+    valid_regex_sequence = re.compile(r'''
+        (\.?[^./ ])([^/]*[^./ ])? # match paths not followed or preceeded by '/', ' ', '.' zero or more times
+        (/(\.?[^./ ])([^/]*[^./ ])?)* # except for '/', ' ', '.' being in between other sequences not at boundaries
+    ''', re.X)
+
+    disallowed_path_components = {
+        'CON', 'PRN', 'AUX', 'NUL',
+        *(f'{cmd}{i}' for cmd in ['COM', 'LPT'] for i in range(1, 10))
+    }
+
+    @classmethod
+    def _sanitize_path(cls, path: str) -> str:
+        """
+        >>> f = CurlManifestGenerator._sanitize_path
+        >>> f('foo/bar/\\x1F/file') # doctest: +NORMALIZE_WHITESPACE
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Invalid file name', 'foo/bar/\\x1f/file', '\\\\x00 to \\\\x1F and `\\\\` are invalid
+        characters')
+
+        >>> f('foo/bar/\\/file') # doctest: +NORMALIZE_WHITESPACE
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Invalid file name', 'foo/bar/\\\\/file', '\\\\x00 to \\\\x1F and `\\\\` are invalid
+        characters')
+
+        >>> f('foo/bar/COM6/file')
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Invalid path component', {'COM6'})
+
+        >>> f('foo/bar/ / baz/file')
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Invalid name sequence in file name', 'foo/bar/ / baz/file')
+
+        >>> f('foo/bar?/fi*le|')
+        'foo/bar_/fi_le_'
+
+        >>> f('foo/bar/file.fastqgz')
+        'foo/bar/file.fastqgz'
+        """
+
+        require(re.search(r'[\x00-\x1f\\]', path) is None,
+                'Invalid file name',
+                path,
+                '\\x00 to \\x1F and `\\` are invalid characters')
```

Comment:
Technically, they are not, they're just not allowed to occur in these paths. Since these messages reach the end user, you don't want to be too technical with the syntax.

```suggestion
                'Control characters and backslash are not allowed.', match.start())
```

which means that you have to extract a `match` variable and cannot use reject or require anymore.

Reply:
I changed the end of your suggestion `match.start()` which returns the index of the offending character to `path[match.start()]` so that the disallowed character is shown.

Reply:
Nope. Think: why would I want to log the position instead of the character itself?